src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializer.cs

```diff
@@ -348,16 +417,365 @@ public virtual object CreateResourceInstance(IEdmStructuredTypeReference structu
+            // in which case we fall back to building the path based on the current path and segments.
+            if (resourceWrapper.ResourceBase.Id != null)
+            {
+                try
+                {
+                    IWebApiRequestMessage internalRequest = readContext.InternalRequest;
+                    IWebApiUrlHelper urlHelper = readContext.InternalUrlHelper;
+
+                    string serviceRoot = urlHelper.CreateODataLink(
+                        internalRequest.Context.RouteName,
+                        internalRequest.PathHandler,
+                        new List<ODataPathSegment>());
+
+                    ODataUriParser parser = new ODataUriParser(readContext.Model, new Uri(serviceRoot), resourceWrapper.ResourceBase.Id);
+
+                    ODataPath odataPath = parser.ParsePath();
+                    if (odataPath != null)
+                    {
+                        return new Routing.ODataPath(odataPath);
+                    }
+                }
+                catch (Exception)
+                {
+                };
+            }
+
+            Routing.ODataPath path = readContext.Path;
+
+            if (path == null)
+            {
+                return null;
+            }
+
+            IEdmEntityType entityType = path.EdmType.AsElementType() as IEdmEntityType;
+
+            if (entityType != null && path.EdmType.TypeKind == EdmTypeKind.Collection)
+            {
+                // create the uri for the current object, using path and key values
+                List<KeyValuePair<string, object>> keys = new List<KeyValuePair<string, object>>();
+                foreach (IEdmStructuralProperty keyProperty in entityType.Key())
+                {
+                    string keyName = keyProperty.Name;
+                    ODataProperty property = resourceWrapper.ResourceBase.Properties.Where(p => p.Name == keyName).FirstOrDefault();
+                    if (property == null && !readContext.DisableCaseInsensitiveRequestPropertyBinding)
+                    {
+                        //try case insensitive
+                        List<ODataProperty> candidates = resourceWrapper.ResourceBase.Properties.Where(p => String.Equals(p.Name, keyName, StringComparison.InvariantCultureIgnoreCase)).ToList();
+                        property = candidates.Count == 1 ? candidates.First() : null;
+                    }
+
+                    object keyValue = property?.Value;
+                    if (keyValue == null)
+                    {
+                        // Note: may be null if the payload did not include key values,
+                        // but still need to add the key so the path is semantically correct.
+                        // Key value type is not validated, so just use string.
+                        // Consider adding tests to ODL to ensure we don't validate key property type in future.
+                        keyValue = "Null";
```

Comment:
Curious about the effects (if any) of doing this...

Reply:
it just for making the path to be semantically correct